lib/horde/supervisor.ex

```diff
@@ -0,0 +1,423 @@
+
+      state.processes
+      |> Enum.filter(fn
+        {_id, {^this_node, _child_spec}} -> true
+        _ -> false
+      end)
+      |> Enum.map(fn {id, {_this_node, child_spec}} ->
+        Task.async(fn ->
+          # shut down the child, remove from the supervisor
+          :ok = Supervisor.terminate_child(state.supervisor_pid, id)
+          :ok = Supervisor.delete_child(state.supervisor_pid, id)
+
+          # mark child as unassigned in the CRDT
+          GenServer.cast(
+            state.processes_pid,
+            {:operation, {@crdt, :add, [child_spec.id, {nil, child_spec}]}}
+          )
+        end)
+      end)
+      |> Enum.map(fn task -> Task.await(task, @shutdown_wait) end)
+
+      # allow time for state to propagate normally to other nodes
+      Process.sleep(10000)
+
+      :ok = Supervisor.stop(state.supervisor_pid)
+
+      send(horde, :force_shutdown)
+    end)
+  end
+
+  defp handle_updated_members_pids(state, new_state) do
+    new_pids =
+      Enum.map(new_state.members, fn {_key, {_, _, _, m, _}} -> m end) |> Enum.into(MapSet.new())
+
+    old_pids =
+      Enum.map(state.members, fn {_node_id, {_, _, _, m, _}} -> m end) |> Enum.into(MapSet.new())
+
+    # if there are any new pids in `member_pids`
+    if MapSet.difference(new_pids, old_pids) |> Enum.any?() do
+      send(state.members_pid, {:add_neighbours, new_pids})
+      send(state.members_pid, :ship_interval_or_state_to_all)
+    end
+  end
+
+  defp handle_updated_process_pids(state, new_state) do
+    new_pids =
+      Enum.map(new_state.members, fn {_key, {_, _, _, _, p}} -> p end) |> Enum.into(MapSet.new())
+
+    old_pids =
+      Enum.map(state.members, fn {_node_id, {_, _, _, _, p}} -> p end) |> Enum.into(MapSet.new())
+
+    if MapSet.difference(new_pids, old_pids) |> Enum.any?() do
+      send(state.processes_pid, {:add_neighbours, new_pids})
+      send(state.processes_pid, :ship_interval_or_state_to_all)
+    end
+  end
+
+  defp dead_members(%{members: members}) do
+    Enum.filter(members, fn
+      {_, {:dead, _, _, _, _}} -> true
+      _ -> false
+    end)
+    |> Enum.map(fn {node_id, _} -> node_id end)
+  end
+
+  defp handle_topology_changes(state) do
+    this_node_id = state.node_id
+
+    dead_members(state)
+    |> Enum.map(fn dead_node ->
+      state.processes
+      |> Enum.filter(fn
+        {_id, {^dead_node, _child_spec}} -> true
+        _ -> false
+      end)
+      |> Enum.filter(fn {_id, {_node, child}} ->
+        case choose_node(child.id, state) do
+          {^this_node_id, _node_info} -> true
+          _ -> false
+        end
+      end)
+      |> Enum.map(fn {_id, {_node, child}} -> add_child(child, state) end)
+    end)
+  end
+
+  defp claim_unclaimed_processes(state) do
+    this_node_id = state.node_id
+
+    state.processes
+    |> Enum.each(fn
+      {id, {nil, child_spec}} ->
+        case choose_node(id, state) do
+          {^this_node_id, _node_info} -> add_child(child_spec, state)
```

Comment:
both this and previous line call into choose_node/2, will this make any impact on performance?

Reply:
The biggest performance problems right now are in the CRDT, I'll be focussing on that first.